### PR TITLE
`hamiltonian()` and `operation()` methods to convert `PauliSentence` into native PL operators or Hamiltonians (PauliUtility #3)

### PR DESCRIPTION
**Pull request by Jaybsoni**
**Context:**
Introduce the `PauliSentence.hamiltonian()` and `PauliSentence.operation()` methods which allow users to go from the lightweight representation of Pauli sentences to native PennyLane operations. 

**Description of the Change:**
- Introduce `PauliWord.hamiltonian()` and `PauliWord.operation()` methods + tests
- Introduce `PauliSentence.hamiltonian()` and `PauliSentence.operation()` methods + tests

**Benefits:**
- More interfacing support between this representation and PL

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -45,7 +45,25 @@
   ```
 
 * New basis sets, `6-311g` and `CC-PVDZ`, are added to the qchem basis set repo.
-  [#3279](https://github.com/PennyLaneAI/pennylane/pull/3279)
+  [(#3279)](https://github.com/PennyLaneAI/pennylane/pull/3279)
```
Comment (AlbertMitjans) [inline doc/releases/changelog-dev.md:48]: We should stick with one and change it everywhere in the file

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -45,7 +45,25 @@
...
+  
+* Added two new methods `operation()`, `hamiltonian()` for both to generate an equivalent PennyLane
```
Comment (AlbertMitjans) [inline doc/releases/changelog-dev.md:50]:
```suggestion
* Added two new methods `operation()`, `hamiltonian()` for both `PauliSentence` and `PauliWord` classes to generate an equivalent PennyLane 
```

File: pennylane/ops/op_math/exp.py [hunk 3]
```diff
@@ -171,6 +170,8 @@ def _queue_category(self):
     # pylint: disable=arguments-renamed, invalid-overridden-method
     @property
     def has_decomposition(self):
+        from pennylane.pauli import is_pauli_word  # pylint: disable=import-outside-toplevel
```
Comment (AlbertMitjans) [inline pennylane/ops/op_math/exp.py:173]: Why? If there is a circular import I would prefer using `qml.pauli.is_pauli_word` instead

File: pennylane/ops/op_math/exp.py [hunk 4]
```diff
@@ -193,6 +194,8 @@ def decomposition(self):
         Returns:
             list[PauliRot]: decomposition of the operator
         """
+        from pennylane.pauli import pauli_word_to_string  # pylint: disable=import-outside-toplevel
```
Comment (AlbertMitjans) [inline pennylane/ops/op_math/exp.py:197]: Same here

File: pennylane/pauli/pauli_arithmetic.py [hunk 5]
```diff
@@ -186,6 +203,34 @@ def to_mat(self, wire_order, format="dense"):
...
+            if wire_order in (None, [], wires.Wires([])):
+                raise ValueError("Can't get the operation for an empty PauliWord.")
+            return Identity(wires=wire_order)
+
+        if len(self) == 1:
+            (wire,) = self
+            op = self[wire]
+            return op_map[op](wire)
+
+        return prod(*(op_map[op](wire) for wire, op in self.items()))
```
Comment (AlbertMitjans) [inline pennylane/pauli/pauli_arithmetic.py:218]: Just a suggestion:
```suggestion
        factors = [op_map[op](wire) for wire, op in self.items()]
        return factors[0] if len(factors) == 1 else prod(*factors)
```

File: pennylane/pauli/pauli_arithmetic.py [hunk 6]
```diff
@@ -186,6 +203,34 @@ def to_mat(self, wire_order, format="dense"):
...
+            if wire_order in (None, [], wires.Wires([])):
+                raise ValueError("Can't get the Hamiltonian for an empty PauliWord.")
+            return Hamiltonian([1], [Identity(wires=wire_order)])
+
+        if len(self) == 1:
+            (wire,) = self
+            op = self[wire]
+            return Hamiltonian([1], [op_map[op](wire)])
+
+        return Hamiltonian([1], [Tensor(*(op_map[op](wire) for wire, op in self.items()))])
```
Comment (AlbertMitjans) [inline pennylane/pauli/pauli_arithmetic.py:232]: Same suggestion:
```suggestion
        obs = [op_map[op](wire) for wire, op in self.items()]
        return Hamiltonian([1], [obs[0] if len(obs) == 1 else Tensor(*obs)])
```

File: pennylane/pauli/pauli_arithmetic.py [hunk 7]
```diff
@@ -280,6 +336,54 @@ def to_mat(self, wire_order, format="dense"):
...
+            )
+
+        return op_sum(
+            *(
+                s_prod(coeff, pw.operation())
+                if len(pw) > 0
+                else s_prod(coeff, pw.operation(wire_order=list(self.wires)))
+                for pw, coeff in self.items()
+            )
+        )
```
Comment (AlbertMitjans) [inline pennylane/pauli/pauli_arithmetic.py:362]: Same suggestion:
```suggestion
        summands = [
            s_prod(coeff, pw.operation(wire_order=list(self.wires))) for pw, coeff in self.items()
        ]
        return summands[0] if len(summands) == 1 else op_sum(*summands)
```

File: pennylane/pauli/pauli_arithmetic.py [hunk 8]
```diff
@@ -280,6 +336,54 @@ def to_mat(self, wire_order, format="dense"):
...
+                if len(pw) > 0
+                else coeff * pw.hamiltonian(wire_order=list(self.wires))
+            )
+
+        return sum(
+            coeff * pw.hamiltonian()
+            if len(pw) > 0
+            else coeff * pw.hamiltonian(wire_order=list(self.wires))
+            for pw, coeff in self.items()
+        )
```
Comment (AlbertMitjans) [inline pennylane/pauli/pauli_arithmetic.py:385]: Same suggestion:
```suggestion
        return sum(
            coeff * pw.hamiltonian(wire_order=list(self.wires)) for pw, coeff in self.items()
        )
```

**Review by AlbertMitjans — COMMENTED** [review]
Good job! Left some comments and some suggestions

File: pennylane/ops/op_math/exp.py (hunk 3 above)
Comment (Jaybsoni) [reply pennylane/ops/op_math/exp.py:173]: I tried it with that syntax first and it still throws the circular import error

File: pennylane/ops/op_math/exp.py (hunk 4 above)
Comment (Jaybsoni) [reply pennylane/ops/op_math/exp.py:197]: I tried it with that syntax first and it still throws the circular import error

File: doc/releases/changelog-dev.md (hunk 1 above)
Comment (Jaybsoni) [reply doc/releases/changelog-dev.md:48]: I think this is the more standard notation,

File: pennylane/ops/op_math/exp.py (hunk 3 above)
Comment (AlbertMitjans) [reply pennylane/ops/op_math/exp.py:173]: Are you sure? I just tried it locally and it works

File: doc/releases/changelog-dev.md (hunk 1 above)
Comment (AlbertMitjans) [reply doc/releases/changelog-dev.md:48]: Could you change it everywhere in the file? This way we will avoid confusions.

**Comment by Jaybsoni** [comment]
[sc-28974]

File: doc/releases/changelog-dev.md (hunk 1 above)
Comment (Jaybsoni) [reply doc/releases/changelog-dev.md:94]: Added 👍🏼

File: pennylane/ops/op_math/exp.py (hunk 3 above)
Comment (Jaybsoni) [reply pennylane/ops/op_math/exp.py]: Trying again see commit [600663d](https://github.com/PennyLaneAI/pennylane/pull/3391/commits/600663dc07aceea6aa8f73cd4dcc3a530476e8f6)

Comment (Jaybsoni) [reply pennylane/ops/op_math/exp.py]: Ahh nice catch it seems to work! I guess it failed for a different reason last time I tried it 🤔

File: pennylane/pauli/pauli_arithmetic.py [hunk 9]
```diff
@@ -257,19 +294,30 @@ def to_mat(self, wire_order, format="dense"):
...
+            """To account for empty pauli_words which represent identity operations."""
+            if w:
+                return wires.Wires(w)
+
+            ps_wires = self.wires
+            if len(ps_wires) > 0:
+                return wires.Wires(
+                    list(ps_wires)[0]
+                )  # return any wire from the Pauli sentence's wires
+            return wires.Wires([])
```
Comment (AlbertMitjans) [inline pennylane/pauli/pauli_arithmetic.py:308]:
```suggestion
            return wires.Wires(self.wires[0]) if len(self.wires) > 0 else wires.Wires([])
```

Comment (AlbertMitjans) [inline pennylane/pauli/pauli_arithmetic.py:308]:
```suggestion
            return w or (wires.Wires(self.wires[0]) if len(self.wires) > 0 else wires.Wires([]))
```

Comment (AlbertMitjans) [reply pennylane/pauli/pauli_arithmetic.py:308]: Suggestion from sourcery 🤯

**Review by AlbertMitjans — APPROVED** [review]
Approved! But be careful because [this PR](https://github.com/PennyLaneAI/pennylane/pull/3384/files) is also fixing the empty Pauli word bug. Make sure that the correct version is pushed to master.

**Pull request merged**